### PR TITLE
client: fix the pd client could be blocked in some cases (#3283)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #3283 to release-4.0

---

Signed-off-by: leoppro <zhaoyilin@pingcap.com>

### What problem does this PR solve?

Fix the PD client could be blocked in some cases

### What is changed and how it works?

There are two levels of context in PD client: `client context` and `tso request context`.
the `tsoRequest.Wait()` function could be blocked when the `client context` is canceled because the `Wait()` function haven't listened the `client context`

### Benchmark
the version of PD server: v4.0.8

before this PR:

```
Start benchmark #0, duration: 5s
Create 3 client(s) for benchmark
count:768290, max:9, min:0, avg:0, >1ms:34372, >2ms:2037, >5ms:457, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:800952, max:9, min:0, avg:0, >1ms:96750, >2ms:13593, >5ms:1495, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:745906, max:8, min:0, avg:0, >1ms:12877, >2ms:3652, >5ms:1330, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:759058, max:4, min:0, avg:0, >1ms:11232, >2ms:5776, >5ms:0, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0

Total:
count:3074206, max:9, min:0, avg:0, >1ms:155231, >2ms:25058, >5ms:3282, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:3074206, >1ms:5.05%, >2ms:0.82%, >5ms:0.11%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%
```

after this PR:

```
Start benchmark #0, duration: 5s
Create 3 client(s) for benchmark
count:1361814, max:7, min:0, avg:1, >1ms:900605, >2ms:204456, >5ms:2896, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:1357623, max:11, min:0, avg:1, >1ms:931611, >2ms:151281, >5ms:3936, >10ms:1, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:737217, max:7, min:0, avg:0, >1ms:3830, >2ms:3444, >5ms:1028, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:1033124, max:8, min:0, avg:0, >1ms:430725, >2ms:83313, >5ms:460, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0

Total:
count:4489778, max:11, min:0, avg:1, >1ms:2266771, >2ms:442494, >5ms:8320, >10ms:1, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:4489778, >1ms:50.49%, >2ms:9.86%, >5ms:0.19%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%
```

### Check List

Tests

- Unit test

### Release note

- Fix the pd client could be blocked in some cases

### CODE REVIEW

**Comment by JmPotato** [comment]
@leoppro Please resolve the conflicts.

File: client/client.go [hunk 1]
```diff
@@ -122,9 +122,22 @@ func WithGroup(group string) ScatterRegionOption {
...
+	physical   int64
+	logical    int64
+	dcLocation string
+}
+
+type lastTSO struct {
+>>>>>>> 52fa5093... client: fix the pd client could be blocked in some cases (#3283)
 	physical int64
 	logical  int64
 }
```
Comment (leoppro) [inline client/client.go:143]:
```suggestion
type tsoRequest struct {
	start    time.Time
	clientCtx  context.Context
	requestCtx context.Context
	done     chan error
	physical int64
	logical  int64
}
```

File: client/client.go [hunk 2]
```diff
@@ -439,6 +452,11 @@ func (c *client) GetTSAsync(ctx context.Context) TSFuture {
...
 	}
 	req := tsoReqPool.Get().(*tsoRequest)
+<<<<<<< HEAD
+=======
+	req.requestCtx = ctx
+	req.clientCtx = c.ctx
+>>>>>>> 52fa5093... client: fix the pd client could be blocked in some cases (#3283)
 	req.start = time.Now()
 	req.ctx = ctx
 	req.physical = 0
```
Comment (leoppro) [inline client/client.go:462]:
```suggestion
	req.requestCtx = ctx
	req.clientCtx = c.ctx
	req.start = time.Now()
	req.physical = 0
```

**Comment by JmPotato** [comment]
@leoppro It seems that applying suggestions doesn't have sign-off and misses fomating to the codes.

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**